Fixed #32129 -- Updated method descriptions in sessions docs.

sarahboyce opened the pull request:
Hello :wave: my first attempt at a ticket, quite excited

Updated the method descriptions of `get_session_cookie_age` `get_expiry_age` `get_expiry_date` in the docs, as well as the method doc strings, to be more explicit and hopefully less misleading.
This is based off this ticket: https://code.djangoproject.com/ticket/32129 which referenced some times members of the community were confused.

I hope this is better but happy to have any feedback and suggestions

carltongibson reviewed (commented):
Hi @sarahboyce 👋

> Hello 👋 my first attempt at a ticket, quite excited

We're excited to have you! Welcome aboard! ⛵️

Thanks for taking this on. It's tricky. The code is very old, and not really 
(I think) intended to be used in the way that folks are wanting to _retrieve_ 
the session expiry. 

A comment from ticket-18458: 

> I recently worked on this code and came to the conclusion that get_expiry_age and get_expiry_date are designed to work at the time when the session is saved, and not at any other point in time.

The related commit there cd17a24083f3ef17cf4c40a41c9d03c250d817c6 which adds
the `modification` parameter needed to retrieve the expiry date has: 

> This change allows for cleaner tests: we can test the exact output.

i.e. it was at least in part a testing convenience (although used to invalidate file backed sessions too.)

Nonetheless, if you have the correct `modification` you **can** get the
remaining time out of these functions so if we can clarify what's going on
that'd be great. 

I've left one inline comment for the proposal on the docstring for `get_expiry_age()`
Have a look at that. Let me know what you think. 

Are you [set up with the test suite](https://docs.djangoproject.com/en/4.0/internals/contributing/writing-code/unit-tests/)? Running only the sessions tests...

```
$ ./runtests.py sessions_tests.tests.DatabaseSessionTests --parallel=1 --pdb
```

... and dropping in a breakpoint, and experimenting with changes, will let you
see the behaviours. Do you think we have tests for each of the use-case there?
If not perhaps an additional test or two helps spell-out what's going on. (If
so they help see each code path in action.)

For the docs, I think I'd say something like:

- Used by the session backends to determine session expiry when saving the session. 
- An explicit expiry can be passed, to be used rather a previous set (or default) expiry. 
- If provided the correct modification, can be used to calculate the remaining session time, rather than the maximum session lifetime. 

That needs phrasing work... 

Overall, perhaps an admonition steering people away from this API would be better: assuming I've got `modification` in hand (which I need for this use-case) I think I'd just do:

```
expires_at = modification + timedelta(seconds=settings.SESSION_COOKIE_AGE)
```

... and not use these methods at all. (Loops back to comment from ticket-18458 at the top...) 

What do you think? 
Thanks again! And welcome again! 😃

sarahboyce commented:
Hi @carltongibson :wave: lovely to meet you, thank you for the thoughtful review!

I've tried to incorporate all the feedback. The context as to when these methods were designed to be used is very helpful.

### Test cases
Got them working :tada: I may have gone overkill here, but I believe each combination of the sessions set expiry vs passing in the `expiry` and using vs not using `modification` are now covered. 
As these tests are on Mixin `SessionTestsMixin` for different session backends I had some issues with `CustomDatabaseSessionTests` as this is testing a backend that had overwritten `get_session_cookie_age` to no longer be the setting value. I have very unimaginately overwritten the tests on this.
Happy to trim it down if a lot of it is not necessary.

### Docstrings
Hopefully these are now in the format with the summary line and further details underneath

### Docs
I played with a warning and then decided against. I have gone pretty much with what you've suggested.
The main difference I _believe_ I have found is that
although knowing `modification` does seem to then give you the correct expiry date for `get_expiry_date`
For `get_expiry_age` this is only true if the expiry has been explicitly set as a date (this also happens when you set to a timedelta as in the background it is set to a date). I'm not sure if that is a bug and an element of the difference between now and the modification should be incorporated in the other cases or if, as you've mentioned, it is just designed to be used on session save so this behaviour is fine.

carltongibson commented on docs/topics/http/sessions.txt:
```diff
@@ -275,26 +275,29 @@ You can edit it multiple times.
+      An explicit ``expiry`` can be passed as an optional keyword argument,
+      to be used rather a previous set (or default) expiry.
 
-      - ``modification``: last modification of the session, as a
-        :class:`~datetime.datetime` object. Defaults to the current time.
-      - ``expiry``: expiry information for the session, as a
-        :class:`~datetime.datetime` object, an :class:`int` (in seconds), or
-        ``None``. Defaults to the value stored in the session by
-        :meth:`set_expiry`, if there is one, or ``None``.
+      If provided the correct ``modification`` (a :class:`~datetime.datetime`
+      object), ``get_expiry_age`` can be used to calculate the remaining
+      session time **only** when the session expiry has been set to an explicit
+      date, in all other cases returns the maximum session lifetime.
```
Going over this again and again, I still lean towards just steering folks away with this kind of thing: 

```diff
diff --git a/docs/topics/http/sessions.txt b/docs/topics/http/sessions.txt
index c36554f22c..40b29d4e4a 100644
--- a/docs/topics/http/sessions.txt
+++ b/docs/topics/http/sessions.txt
@@ -288,13 +288,32 @@ You can edit it multiple times.
         ``None``. Defaults to the value stored in the session by
         :meth:`set_expiry`, if there is one, or ``None``.
 
+      .. note::
+
+        This method is used by session backends to determine the session expiry
+        age in seconds when saving the session. It is not really intended for
+        usage outside of that context.
+
+        In particular, whilst it is **possible** to determine the remaining
+        lifetime of a session **just when** you have the correct
+        ``modification`` value **and** the ``expiry`` is set as a ``datetime``
+        object, where you do have the ``modification``, it is more
+        straight-forward to calculate the expiry by-hand::
+
+            expires_at = modification + timedelta(seconds=settings.SESSION_COOKIE_AGE)
+
+        In all cases where ``expiry`` is something other than a ``datetime``
+        object, ``get_expiry_age()`` returns the maximum session lifetime,
+        which is likely not what you intend.
+
     .. method:: get_expiry_date()
 
       Returns the date this session will expire. For sessions with no custom
       expiration (or those set to expire at browser close), this will equal the
       date :setting:`SESSION_COOKIE_AGE` seconds from now.
 
-      This function accepts the same keyword arguments as :meth:`get_expiry_age`.
+      This function accepts the same keyword arguments as
+      :meth:`get_expiry_age`, and similar notes on usage apply.
 
     .. method:: get_expire_at_browser_close()
```

diff from current `main`. 

Anyone implementing a session backend can go and look at the source. For everyone else it feels like too much info, when what we really want to say is _Don't use this_. 

What do you think?

sarahboyce replied:
Yeah I think that directs the users better, rather the warning should help direct them away.

With each of the original descriptions I added "on session save" in the first sentences. Which I hope is not more confusing. It's all better clarified in the note.
I'm becoming doc blind :joy: but I think we're getting there.

carltongibson commented on docs/topics/http/sessions.txt:
```diff
@@ -245,8 +245,8 @@ You can edit it multiple times.
 
-      Returns the age of session cookies, in seconds. Defaults to
-      :setting:`SESSION_COOKIE_AGE`.
+      Returns the value of the setting :setting:`SESSION_COOKIE_AGE`
+      by default. This method can be overwritten by the session backend.
```
I think this change is right. 👍 

```suggestion
      Returns the value of the setting :setting:`SESSION_COOKIE_AGE`,
      by default. This method can be overwritten by the session backend.
```

carltongibson reviewed (commented):
Hi @sarahboyce — thanks for the quick follow-up! 

A couple of points. Let me know what you think about both: 

* My main comment is on the docs `sessions.txt` — see the inline comment there. 
* I'm not sure we need the docstring changes in `base.py` — the functions aren't long. If I'm working on a backend I can just see them. (The confusion in the ticket is for an end-user trying to calculate the **remaining** session lifetime — which without a `modification` they can't do. I think the docs change is the focus of that.)

Then I didn't quite make it to the tests again. (Let's get the other changes right and then we can decide.) 

Q: Are you able to rebase on latest `main` and squash down to a single commit?

sarahboyce commented:
@carltongibson squashed and rebased :+1: I can keep doing this as we tweak 
Removed the change to the doc string, only changes to the docs and tests remain.
Tests updates were a good learning exercise so not bothered around dropping the changes if they are not necessary.

carltongibson commented on docs/topics/http/sessions.txt:
```diff
@@ -275,26 +275,36 @@ You can edit it multiple times.
+        usage outside of that context.
 
-      This function accepts two optional keyword arguments:
+        In particular, whilst it is **possible** to determine the remaining
```
Accursed `en-us` 😃 Thanks @smithdc1
